flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java

```diff
@@ -513,36 +514,39 @@ private void stopDispatcherServices() throws Exception {
-
-        try {
-            if (isInGloballyTerminalState(jobID)) {
-                log.warn(
-                        "Ignoring JobGraph submission '{}' ({}) because the job already reached a globally-terminal state (i.e. {}) in a previous execution.",
-                        jobGraph.getName(),
-                        jobID,
-                        Arrays.stream(JobStatus.values())
-                                .filter(JobStatus::isGloballyTerminalState)
-                                .map(JobStatus::name)
-                                .collect(Collectors.joining(", ")));
-                return FutureUtils.completedExceptionally(
-                        DuplicateJobSubmissionException.ofGloballyTerminated(jobID));
-            } else if (jobManagerRunnerRegistry.isRegistered(jobID)
-                    || submittedAndWaitingTerminationJobIDs.contains(jobID)) {
-                // job with the given jobID is not terminated, yet
-                return FutureUtils.completedExceptionally(
-                        DuplicateJobSubmissionException.of(jobID));
-            } else if (isPartialResourceConfigured(jobGraph)) {
-                return FutureUtils.completedExceptionally(
-                        new JobSubmissionException(
-                                jobID,
-                                "Currently jobs is not supported if parts of the vertices have "
-                                        + "resources configured. The limitation will be removed in future versions."));
-            } else {
-                return internalSubmitJob(jobGraph);
-            }
-        } catch (FlinkException e) {
-            return FutureUtils.completedExceptionally(e);
-        }
+        return isInGloballyTerminalState(jobID)
+                .thenCompose(
+                        isTerminated -> {
+                            if (isTerminated) {
+                                log.warn(
+                                        "Ignoring JobGraph submission '{}' ({}) because the job already "
+                                                + "reached a globally-terminal state (i.e. {}) in a "
+                                                + "previous execution.",
+                                        jobGraph.getName(),
+                                        jobID,
+                                        Arrays.stream(JobStatus.values())
+                                                .filter(JobStatus::isGloballyTerminalState)
+                                                .map(JobStatus::name)
+                                                .collect(Collectors.joining(", ")));
+                                return FutureUtils.completedExceptionally(
+                                        DuplicateJobSubmissionException.ofGloballyTerminated(
+                                                jobID));
+                            } else if (jobManagerRunnerRegistry.isRegistered(jobID)
```

Comment:
```suggestion
                            } else if (jobManagerRunnerRegistry.isRegistered(jobID)
```
Ok, that's a tricky one which I missed: The Dispatcher has one requirement: Any state access needs to happen in the main thread of the Dispatcher. There's a special implementation of `JobManagerRunnerRegistry` that ensures this invariant (see [OnMainThreadJobmanagerRunnerRegistry](https://github.com/apache/flink/blob/3efd4c2cf1be670f499e6637445e283e48deee60/flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/OnMainThreadJobManagerRunnerRegistry.java)). Our change (with the `thenCompose` being chained with the future that's returned by `isInGloballyTerminalState`) is going against this requirement. Why?

`isInGloballyTerminalState` calls `jobResultStore.hasJobResultEntryAsync` which executes the logic on the `ioExecutor` (i.e. a thread for IO operations which is not the Dispatcher's main thread) internally. The returned future is linked to this executor, i.e. any chained `CompletableFuture` calls will run in the same thread. The `thenCompose` logic is, therefore, also executed in the `ioExecutor` instead of the main thread.

To workaround this, we have to change the executor for the chained execution. This can be achieved by using `thenComposeAsync`, instead. Here we would specify the main thread executor by calling `getMainThreadExecutor`. One example where it's done like that is [Dispatcher:619](https://github.com/apache/flink/blob/b6992b9d80e8ca9c6a81d198d9ed628821e4ab49/flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java#L619): The `cleanupAsync` method is executed on the `ioExecutor`. But the error handling has to happen in the main thread again. That's where we use `handleAsync` with `getMainThreadExecutor`.

Reply:
Thanks for your detailed explanation! 😄 I've modified the `thenCompose` to  `thenComposeAsync`.

Reply:
You didn't do a pull before adding the changes (I did a [force-push](https://github.com/apache/flink/pull/22341#pullrequestreview-1588695382) to include a few minor changes previous). These changes were reverted with your most-recent push

Reply:
😼 I have added the changes in the [comparison](https://github.com/apache/flink/compare/5a7329d02ce51f53bfd4ab8c2f6e30b5271c539f..ee05e69244ad3254c05a567e37e6be7d85ce0881). Really sorry for this.

Reply:
No worries. Let's see whether CI becomes :green_circle: this time. :crossed_fingers: